cloudflare/src/request.rs

```diff
@@ -0,0 +1,87 @@
+use crate::{Credentials, Endpoint, CLIENT_API_V4_URI};
+use url::Url;
+
+/// A basic builder that contains credentials and a base API URL for use with constructing requests
+/// for Endpoints.
+pub struct RequestBuilder {
+    /// The credentials to use for requests
+    pub credentials: Credentials,
+    /// The API URL to make requests to
+    pub api_url: Url,
+}
+
+impl RequestBuilder {
+    /// Creates a new builder with the provided credentials and the default client API url.
+    pub fn new(credentials: Credentials) -> Self {
+        Self {
+            credentials,
+            api_url: CLIENT_API_V4_URI.parse().expect("default API URI is valid"),
+        }
+    }
+
+    /// Attempts to create a request using this builder and the specified endpoint. If the API Url
+    /// cannot be joined with the endpoint's path, this returns a URL parse error.
+    pub fn try_build<'a, E: Endpoint + 'a>(
+        &'a self,
+        endpoint: &'a E,
+    ) -> Result<Request<'a, E>, url::ParseError> {
+        Ok(Request {
+            url: self.api_url.join(&endpoint.path())?,
+            credentials: &self.credentials,
+            query: endpoint.query(),
+            body: endpoint.body(),
+        })
+    }
+
+    /// Creates a request using this builder and the specified endpoint.
+    ///
+    /// # Panics
+    ///
+    /// If the API Url cannot be joined with the endpoint's path.
```

Comment:
weird to see API capitalized but URL not, we should choose one policy and be consistent

Reply:
I think I made it not capitalized because it refers to the Url type. Should make that obvious with a intra-doc link.